qiita_db/artifact.py

```diff
@@ -0,0 +1,598 @@
+                     VALUES (%s, %s)"""
+            sql_args = [[a_id, fp_id] for fp_id in fp_ids]
+            TRN.add(sql, sql_args, many=True)
+            TRN.execute()
+
+        return instance
+
+    @classmethod
+    def delete(cls, artifact_id):
+        r"""Deletes an artifact from the system
+
+        Parameters
+        ----------
+        artifact_id : int
+            The artifact to be removed
+
+        Raises
+        ------
+        QiitaDBArtifactDeletionError
+            If the artifact is public
+            If the artifact has children
+            If the artifact has been analyzed
+            If the artifact has been submitted to EBI
+            If the artifact has been submitted to VAMPS
+        """
+        with TRN:
+            # This will fail if the artifact with id=artifact_id doesn't exist
+            instance = cls(artifact_id)
+
+            # Check if the artifact is public
+            if instance.visibility == 'public':
+                raise QiitaDBArtifactDeletionError(artifact_id, "it is public")
+
+            # Check if this artifact has any children
+            if instance.children:
+                raise QiitaDBArtifactDeletionError(
+                    artifact_id,
+                    "it has children: %s"
+                    % ', '.join([str(c.id) for c in instance.children]))
+
+            # Check if the artifact has been analyzed
+            sql = """SELECT EXISTS(SELECT *
+                                   FROM qiita.analysis_sample
+                                   WHERE artifact_id = %s)"""
+            TRN.add(sql, [artifact_id])
+            if TRN.execute_fetchlast():
+                raise QiitaDBArtifactDeletionError(
+                    artifact_id, "it has been analyzed")
+
+            # Check if the artifact has been submitted to EBI
+            if instance.can_be_submitted_to_ebi and \
+                    instance.ebi_run_accessions:
+                raise QiitaDBArtifactDeletionError(
+                    artifact_id, "it has been submitted to EBI")
+
+            # Check if the artifact has been submitted to VAMPS
+            if instance.can_be_submitted_to_vamps and \
+                    instance.is_submitted_to_vamps:
+                raise QiitaDBArtifactDeletionError(
+                    artifact_id, "it has been submitted to VAMPS")
+
+            # We can now remove the artifact
+            filepaths = instance.filepaths
+            study = instance.study
+
+            sql = """DELETE FROM qiita.artifact_filepath
+                     WHERE artifact_id = %s"""
+            TRN.add(sql, [artifact_id])
+
+            # If the artifact doesn't have parents, we move the files to the
+            # uploads folder. We also need to nullify the column in the prep
+            # template table
+            if not instance.parents:
+                move_filepaths_to_upload_folder(study.id, filepaths)
+
+                sql = """UPDATE qiita.prep_template
```

Comment:
Rethinking this, I can see that this might make things more difficult in the long run and move_filepaths_to_upload_folder should only move the files ...